cfme/infrastructure/virtual_machines.py

```diff
@@ -101,106 +77,117 @@ def date_retire_element(fill_data):
-                ],
-
-                "vm_templates_orphaned_branch":
-                [
-                    lambda ctx: visible_tree.click_path('<Orphaned>'),
-                    {
-                        "infra_orphan_obj": lambda ctx: visible_tree.click_path(ctx["orphan_name"]),
-                    }
-                ],
-            }
-        ],
-
-        "infra_vms":
-        [
-            lambda _: accordion.tree("VMs", "All VMs"),
-            {
-                "infra_vms_filter_folder":
-                [
-                    lambda ctx: visible_tree.click_path(ctx["folder_name"]),
-                    {
-                        "infra_vms_filter": lambda ctx: visible_tree.click_path(ctx["filter_name"]),
-                    }
-                ],
-
-                "infra_vm_by_name": lambda ctx: sel.click(ctx['vm'].find_quadicon(
-                    do_not_navigate=True))
-            }
-        ],
-
-        "infra_templates":
-        [
-            lambda _: (accordion.tree("Templates", "All Templates"), toolbar.set_vms_grid_view()),
-            {
-                "infra_templates_filter_folder":
-                [
-                    lambda ctx: visible_tree.click_path(ctx["folder_name"]),
-                    {
-                        "infra_templates_filter":
-                        lambda ctx: visible_tree.click_path(ctx["filter_name"]),
-                    }
-                ],
-            }
-        ],
-    }
-)
-
-
-class Common(Pretty):
 
-    pretty_attrs = ["name", "provider_crud", "template_name"]
+##
+# Experimental tree class DSL
+def scavenge_class(cls, ignore_navigate=False):
+    if "navigate" not in cls.__dict__ and not ignore_navigate:
+        raise ValueError(
+            "The nav class {} must contain navigation staticmethod".format(cls.__name__))
+    contents = AttrDict({"subclasses": {}, "direct_navs": {}})
+    for key, value in cls.__dict__.iteritems():
+        if key.startswith("_") or key == "navigate":
+            continue
+        if inspect.isclass(value):
+            contents.subclasses[value.__name__] = scavenge_class(value)
+        elif callable(value) and value.__name__ != "<lambda>":
+            contents.direct_navs[value.__name__] = value
+        # Skipping others
+
+    if not contents.subclasses and not contents.direct_navs and not ignore_navigate:
+        return cls.navigate
+    elif ignore_navigate:
+        result_dict = {}
+        result_dict.update(contents.subclasses)
+        result_dict.update(contents.direct_navs)
+        return result_dict
+    else:
+        result_dict = {}
+        result_dict.update(contents.subclasses)
+        result_dict.update(contents.direct_navs)
+        return [cls.navigate, result_dict]
+
+
+def extend_nav(cls):
+    nav.add_branch(cls.__name__, scavenge_class(cls, ignore_navigate=True))
+
+
+@extend_nav
+class infrastructure_virtual_machines(object):
+    class infra_vm_and_templates(object):
+        @staticmethod
```

Comment:
Yup, I want to do something like that. Will check and if it works, I will incorporate it in the current proposal.